server/test/auth/ApiAuthenticatorTest.java

```diff
@@ -0,0 +1,177 @@
+package auth;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static play.test.Helpers.fakeRequest;
+
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import io.ebean.DB;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Optional;
+import models.ApiKey;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.core.credentials.UsernamePasswordCredentials;
+import org.pac4j.core.exception.BadCredentialsException;
+import org.pac4j.play.PlayWebContext;
+import play.Application;
+import play.ApplicationLoader;
+import play.Environment;
+import play.api.inject.BindingKey;
+import play.cache.NamedCache;
+import play.cache.SyncCacheApi;
+import play.inject.guice.GuiceApplicationBuilder;
+import play.inject.guice.GuiceApplicationLoader;
+import play.mvc.Http;
+import play.test.Helpers;
+import support.ResourceCreator;
+
+public class ApiAuthenticatorTest {
+  @Inject Application application;
+
+  @Inject
+  @NamedCache("api-keys")
+  SyncCacheApi cacheApi;
```

Comment:
Put the Optional.empty() doc here too as it defines the API.

---
Not a blocker, just a comment, but I wonder if we should wrap this in a thin class so that semantic can be in 1 authoritative place, an Inject that instead.

Reply:
moved caching logic into the ApiKeyService with discussion of caching invalid key IDs